src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubListener.cs

```diff
@@ -12,180 +12,198 @@
-                            dispatches.Add(task);
-                        }
+                            TriggerValue = triggerInput.GetSingleEventTriggerInput(i)
+                        };
+                        var invocationTask = _executor.TryExecuteAsync(input, _cts.Token);
+                        invocationTasks.Add(invocationTask);
                     }
 
                     // Drain the whole batch before taking more work
-                    if (dispatches.Count > 0)
+                    if (invocationTasks.Count > 0)
                     {
-                        await Task.WhenAll(dispatches);
+                        await Task.WhenAll(invocationTasks);
                     }
                 }
                 else
                 {
                     // Batch dispatch
-
-                    TriggeredFunctionData input = new TriggeredFunctionData
+                    var input = new TriggeredFunctionData
                     {
-                        ParentId = null,
-                        TriggerValue = value
+                        TriggerValue = triggerInput
                     };
 
-                    FunctionResult result = await this._parent._executor.TryExecuteAsync(input, _cts.Token);
+                    await _executor.TryExecuteAsync(input, _cts.Token);
                 }
 
+                // Dispose all messages to help with memory pressure.
+                // If this is missed, the finalizer thread will still get them. 
                 bool hasEvents = false;
-                // Dispose all messages to help with memory pressure. If this is missed, the finalizer thread will still get them. 
                 foreach (var message in messages)
                 {
                     hasEvents = true;
                     message.Dispose();
                 }
 
-                // Don't checkpoint if no events. This can reset the sequence counter to 0. 
+                // Checkpoint if we procesed any events.
+                // Don't checkpoint if no events. This can reset the sequence counter to 0.
+                // Note: we intentionally checkpoint the batch regardless of function 
+                // success/failure. EventHub doesn't support any sort "poison event" model,
+                // so that is the responsibility of the user's function currently. E.g.
+                // the function should have try/catch handling around all event processing
+                // code, and capture/log/persist failed events, since they won't be retried.
                 if (hasEvents)
                 {
-                    await _checkpoint(context);
+                    await CheckpointAsync(context);
                 }
             }
-        } // end class Listener 
+
+            private async Task CheckpointAsync(PartitionContext context)
+            {
+                if (_config.BatchCheckpointFrequency == 1)
```

Comment:
We're losing some validation we had with the previous implementation to ensure `BatchCheckpointFrequency` wasn't set to 0.

Reply:
Not a huge deal, but this is also slightly less efficient.

Reply:
I removed the validation because if you look at the setter for EventHubConfiguration it already has the validation - it can't be set to an invalid value.

Regarding inefficiency - you're talking about the extra property access? I'll fix that.